File: 필수문항/156_계란으로계란치기.py
import sys

input = sys.stdin.readline

n = int(input())

eggs = []
for _ in range(n):
    eggs.append(list(map(int, input().split())))
ans = 0


# [내구도, 무게]
def dfs(i):
    global eggs, ans, n

    if i == n:
        broken = 0
        for egg in eggs:
            if egg[0] <= 0:
                broken += 1
        ans = max(ans, broken)
        return

    if eggs[i][0] <= 0:
        dfs(i + 1)
        return

    check = False

    for l in range(n):
        if l == i or eggs[l][0] <= 0:
            continue

        check = True
        eggs[i][0] -= eggs[l][1]
        eggs[l][0] -= eggs[i][1]
        dfs(i + 1)
        eggs[i][0] += eggs[l][1]
        eggs[l][0] += eggs[i][1]

    if not check:
        dfs(i + 1)


dfs(0)
print(ans)

File: 필수문항/156_계란으로계란치기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    private static int n;
    private static int[][] eggs;
    private static int ans;

    private static void dfs(int idx) {
        if (idx == n) {
            int broken = 0;
            for (int[] egg : eggs) {
                if (egg[0] <= 0) {
                    broken++;
                }
            }
            ans = Math.max(ans, broken);
        } else if (eggs[idx][0] <= 0) {
            dfs(idx + 1);
        } else {
            boolean check = false;
            for (int i = 0; i < n; i++) {
                if (i == idx) {
                    continue;
                } else if (eggs[i][0] > 0) {
                    check = true;
                    eggs[i][0] -= eggs[idx][1];
                    eggs[idx][0] -= eggs[i][1];
                    dfs(idx + 1);
                    eggs[i][0] += eggs[idx][1];
                    eggs[idx][0] += eggs[i][1];
                }
            }
            if (!check) {
                dfs(idx + 1);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        eggs = new int[n][2];
        StringTokenizer st;
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            eggs[i][0] = Integer.parseInt(st.nextToken());
            eggs[i][1] = Integer.parseInt(st.nextToken());
        }
        dfs(0);
        System.out.println(ans);
    }
}
